crates/node/src/cluster_marker.rs

```diff
@@ -0,0 +1,312 @@
+    IncompatibleClusterName {
+        configured_cluster_name: String,
+        persisted_cluster_name: String,
+    },
+    #[error("current Restate version '{current_version}' is not compatible with data directory. Requiring Restate version >= '{min_version}'")]
+    IncompatibleVersion {
+        current_version: Version,
+        min_version: Version,
+    },
+}
+
+/// Marker which is stored in the Node's working directory telling about the
+/// previous processes that worked on it before. It can be used for sanity checks.
+#[derive(Debug, serde::Serialize, serde::Deserialize, PartialEq, Eq)]
+pub struct ClusterMarker {
+    cluster_name: String,
+    max_version: Version,
+    current_version: Version,
+}
+
+impl ClusterMarker {
+    fn new(cluster_name: String, current_version: Version) -> Self {
+        Self {
+            cluster_name,
+            max_version: current_version.clone(),
+            current_version,
+        }
+    }
+}
+
+pub fn validate_and_update_cluster_marker(
+    cluster_name: &str,
+) -> Result<(), ClusterValidationError> {
+    let current_version = Version::parse(env!("CARGO_PKG_VERSION"))?;
+    let cluster_marker_filepath = node_filepath(CLUSTER_MARKER_FILE_NAME);
+
+    validate_and_update_cluster_marker_inner(
+        cluster_name,
+        current_version,
+        cluster_marker_filepath.as_path(),
+    )
+}
+
+fn validate_and_update_cluster_marker_inner(
+    cluster_name: &str,
+    current_version: Version,
+    cluster_marker_filepath: &Path,
+) -> Result<(), ClusterValidationError> {
+    let tmp_cluster_marker_filepath = cluster_marker_filepath
+        .parent()
+        .expect("filepath should have parent directory")
+        .join(TMP_CLUSTER_MARKER_FILE_NAME);
+    let mut cluster_marker = if cluster_marker_filepath.exists() {
+        let cluster_marker_file = std::fs::File::open(cluster_marker_filepath)
+            .map_err(ClusterValidationError::OpenFile)?;
+        serde_json::from_reader(&cluster_marker_file).map_err(ClusterValidationError::Decode)?
+    } else {
+        info!("Did not find existing cluster marker. Creating a new one.");
+        ClusterMarker::new(cluster_name.to_owned(), current_version.clone())
+    };
+
+    // sanity checks
+    if cluster_marker.cluster_name != cluster_name {
+        return Err(ClusterValidationError::IncompatibleClusterName {
+            configured_cluster_name: cluster_name.to_owned(),
+            persisted_cluster_name: cluster_marker.cluster_name,
+        });
+    }
+
+    // we only support going back one minor version
+    let min_version = Version::new(
+        cluster_marker.max_version.major,
+        cluster_marker.max_version.minor.saturating_sub(1),
+        0,
+    );
+
+    if current_version.cmp_precedence(&min_version) == Ordering::Less {
+        return Err(ClusterValidationError::IncompatibleVersion {
+            current_version,
+            min_version,
+        });
+    }
```

Comment:
I'm not too comfortable putting this as a blanket rule here. Can we disable and only add explicit checks when we release versions that are incompatible with previous ones?

Reply:
We can do this. It adds a manual task to ensure that this information is properly updated at release time, I guess. Do you expect many people to go backward a couple of versions after upgrading?

Reply:
I've added a commit which tries to make the compatibility a bit more flexible by storing some compatibility information in a map.